llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -13710,8 +13710,12 @@ SDValue DAGCombiner::visitZERO_EXTEND(SDNode *N) {
 
   // fold (zext (zext x)) -> (zext x)
   // fold (zext (aext x)) -> (zext x)
-  if (N0.getOpcode() == ISD::ZERO_EXTEND || N0.getOpcode() == ISD::ANY_EXTEND)
-    return DAG.getNode(ISD::ZERO_EXTEND, DL, VT, N0.getOperand(0));
+  if (N0.getOpcode() == ISD::ZERO_EXTEND || N0.getOpcode() == ISD::ANY_EXTEND) {
+    SDNodeFlags Flags;
+    if (N0.getOpcode() == ISD::ZERO_EXTEND)
```

Comment:
Doesn't `zext` always imply nneg unless sizeof(dstty) <= sizeof(srcty) in which case its a nop?

Reply:
nneg is about the sign bit of the input to the zero extend, not the output. It indicates that the zext could be a replaced with a sext.